### DISCUSSION TITLE
v2.0.0-beta.0

### DISCUSSION THREAD

**Discussion by markerikson** [reduxjs/redux-toolkit#3493] [Announcements]
This **beta release** updates many of our TS types for improved type safety and behavior, updates `entityAdapter.getSelectors()` to accept a `createSelector` option, depends on the latest `redux@5.0-beta.0` release, and includes all prior changes from the 2.0 alphas.  This release has **breaking changes**.

```bash
npm i @reduxjs/toolkit@next

yarn add @reduxjs/toolkit@next
```

### Store Configuration Tweaks and Type Safety

We've seen many cases where users passing the `middleware` parameter to `configureStore` have tried spreading the array returned by `getDefaultMiddleware()`, or passed an alternate plain array. This unfortunately loses the exact TS types from the individual middleware, and often causes TS problems down the road (such as `dispatch` being typed as `Dispatch<AnyAction>` and not knowing about thunks).

`getDefaultMiddleware()` already used an internal `MiddlewareArray` class, an `Array` subclass that had strongly typed `.concat/prepend()` methods to correctly capture and retain the middleware types.

We've renamed that type to `Tuple`, and `configureStore`'s TS types now require that you _must_ use `Tuple` if you want to pass your own array of middleware:

```ts
import { configureStore, Tuple } from '@reduxjs/toolkit'

configureStore({
  reducer: rootReducer,
  middleware: new Tuple(additionalMiddleware, logger),
})
```

(Note that this has no effect if you're using RTK with plain JS, and you could still pass a plain array here.)

Similarly, the `enhancers` field used to accept an array directly. It now is a callback that receives a `getDefaultEnhancers` method, equivalent to `getDefaultMiddleware()`:

```ts
const store = configureStore({
  reducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
  enhancers: (getDefaultEnhancers) =>
    getDefaultEnhancers({
      autoBatch: false,
    }).concat(batchedSubscribe(debounceNotify)),
})
```

It too expects a `Tuple` return value if you're using TS.

### Entity Adapter Updates

`entityAdapter.getSelectors()` now accepts an options object as its second argument.  This allows you to pass in your own preferred `createSelector` method, which will be used to memoize the generated selectors.  This could be useful if you want to use [one of Reselect's new alternate memoizers](https://github.com/reduxjs/reselect/releases/tag/v5.0.0-alpha.1), or some other memoization library with an equivalent signature.

`createEntityAdapter` now accepts an `Id` generic argument, which will be used to strongly type the item IDs anywhere those are exposed.

The `.entities` lookup table is now defined to use a standard TS `Record<Id, MyEntityType>`, which assumes that each item lookup exists by default.  Previously, it used a `Dictionary<MyEntityType>` type, which assumed the result was `MyEntityType | undefined`.  The `Dictionary` type has been removed.

If you prefer to assume that the lookups _might_ be undefined, use TypeScript's `noUncheckedIndexedAccess` configuration option to control that.

### New `UnknownAction` Type

The Redux core TS types have always exported an `AnyAction` type, which is defined to have `{type: string}` and treat any other field as `any`. This makes it easy to write uses like `console.log(action.whatever)`, but unfortunately does not provide any meaningful type safety.

We now export an `UnknownAction` type, which treats all fields other than `action.type` as `unknown`.  This encourages users to write type guards that check the action object and assert its _specific_ TS type.  Inside of those checks, you can access a field with better type safety.

`UnknownAction` is now the default any place in the Redux and RTK source that expects an action object.

`AnyAction` still exists for compatibility, but has been marked as deprecated.

Note that [Redux Toolkit's action creators have a `.match()` method](https://redux-toolkit.js.org/api/createAction#actioncreatormatch) that acts as a useful type guard:

```ts
if (todoAdded.match(someUnknownAction)) {
  // action is now typed as a PayloadAction<Todo>
}
```

### Earlier Alpha Changes

Summarizing the changes from earlier alphas:

#### New Features

- `combineSlices` API with built-in support for slice reducer injection for code-splitting
- `selectors` field in `createSlice`
- Callback form of `createSlice.reducers`, which allows defining thunks inside `createSlice`
- "Dynamic middleware" middleware
- `configureStore` adds `autoBatchEnhancer` by default

#### Breaking Changes

- Object argument for `createReducer` and `createSlice.extraReducers` has been removed
- Packaging converted to have full ESM/CJS compatibility
- Dropped UMD build artifacts
- JS build output is now "modern" and not transpiled for IE11 compatibility
- Updated to Immer 10, and dropped the legacy ES5 compat option
- Updated Redux core dep to `5.0-beta`
- `actionCreator.toString()` override removed (although we're reconsidering this)
- Standalone `getDefaultMiddleware` removed
- Other deprecated fields removed

## What's Changed

* Add Id type parameter in createEntityAdapter by @Matt-Ord in https://github.com/reduxjs/redux-toolkit/pull/3187
* Require usage of Tuple in TS by @EskiMojo14 in https://github.com/reduxjs/redux-toolkit/pull/3460
* Require that enhancers is a callback by @EskiMojo14 in https://github.com/reduxjs/redux-toolkit/pull/3461
* fix invalid createEntityAdapter call by @EskiMojo14 in https://github.com/reduxjs/redux-toolkit/pull/3490
* Use Record<EntityId, T> instead of Dictionary<T> by @EskiMojo14 in https://github.com/reduxjs/redux-toolkit/pull/3424
* Fix inference for reducers without actions specified by @EskiMojo14 in https://github.com/reduxjs/redux-toolkit/pull/3475
* Prefer UnknownAction and Action to AnyAction by @EskiMojo14 in https://github.com/reduxjs/redux-toolkit/pull/3363
* Allow passing a `createSelector` instance to adapter.getSelectors by @EskiMojo14 in https://github.com/reduxjs/redux-toolkit/pull/3481

**Full Changelog**: https://github.com/reduxjs/redux-toolkit/compare/v2.0.0-alpha.6...v2.0.0-beta.0

<hr /><em>This discussion was created from the release <a href='https://github.com/reduxjs/redux-toolkit/releases/tag/v2.0.0-beta.0'>v2.0.0-beta.0</a>.</em>

**Comment by einarq** [comment]
RTK 2 is looking really good! Looking forward to the final release.

Small question: Why was Dictionary removed? Found it quite handy. I can of course inline it in the app, but was wondering if it was considered harmful in any way.

**Reply by EskiMojo14** [reply]
see https://github.com/reduxjs/redux-toolkit/issues/930 for the full context - basically the original intention was similar to what noUncheckedIndexedAccess does now. Now that rule exists, we can leave it up to the user how it should behave.

**Reply by ryota-murakami** [reply]
@EskiMojo14 @markerikson 

Great works!
I don't use entityAdapter because I don't facing situation that need normalization.

And one more thing I upgrade RTK v2.0.0 on [this project](https://nsx.malloc.tokyo/) and then I got bunch of type error.
After a quick check, the problem was that I used a Type that was no longer exported from v2 in several places.

.e.g
https://github.com/laststance/nsx/blob/main/src/redux/API.ts#L1

So I would like to confirm with you that in the Redux Toolkit Query, the namespace ["@reduxjs/toolkit/query"](https://github.com/reduxjs/redux-toolkit/blob/master/packages/toolkit/src/query/react/index.ts) Is only correct path for intended to be exposed externally as an API?

**Reply by EskiMojo14** [reply]
@ryota-murakami correct - the only entry points supported are:

- @reduxjs/toolkit
- @reduxjs/toolkit/react
- @reduxjs/toolkit/query
- @reduxjs/toolkit/query/react

Any types you import from /dist or /src that aren't exported from those entry points are technically internal, and thus are liable to change without any specific communication regarding them.

**Reply by ryota-murakami** [reply]
@EskiMojo14  Thanks! I'm going to fix types 😄

**Comment by tommedema** [comment]
When I run `npm i @reduxjs/toolkit@beta` and check package.json it actually installs `"@reduxjs/toolkit": "^1.3.0-beta.1",`

**Reply by markerikson** [reply]
My bad, that should have been `@reduxjs/toolkit@next` - sorry about that!

**Comment by devanecondition** [comment]
I'm trying to add middleware to the default middleware like this (using typescript):

```
export const store = configureStore({
  reducer: {
    // reducers
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(myMiddleware),
})
```

Everything works as expected, however i'm getting a typescript error for `middleware`:

```
Type '(getDefaultMiddleware: GetDefaultMiddleware<any>) => Tuple<[ThunkMiddleware<any, UnknownAction>, Middleware<{}, any, Dispatch<UnknownAction>>]>' is not assignable to type 'Tuple<Middlewares<any>> | ((getDefaultMiddleware: GetDefaultMiddleware<any>) => Tuple<Middlewares<any>>) | undefined'.
  Type '(getDefaultMiddleware: GetDefaultMiddleware<any>) => Tuple<[ThunkMiddleware<any, UnknownAction>, Middleware<{}, any, Dispatch<UnknownAction>>]>' is not assignable to type '(getDefaultMiddleware: GetDefaultMiddleware<any>) => Tuple<Middlewares<any>>'.
    Type 'Tuple<[ThunkMiddleware<any, UnknownAction>, Middleware<{}, any, Dispatch<UnknownAction>>]>' is not assignable to type 'Tuple<Middlewares<any>>'.
      Type '[ThunkMiddleware<any, UnknownAction>, Middleware<{}, any, Dispatch<UnknownAction>>]' is not assignable to type 'Middlewares<any>'.
        Type 'Middleware<{}, any, Dispatch<UnknownAction>> | ThunkMiddleware<any, UnknownAction>' is not assignable to type 'Middleware<{}, any, Dispatch<UnknownAction>>'.
          Type 'ThunkMiddleware<any, UnknownAction>' is not assignable to type 'Middleware<{}, any, Dispatch<UnknownAction>>'.
            Types of parameters 'next' and 'next' are incompatible.
              Type 'unknown' is not assignable to type 'T'.
                'T' could be instantiated with an arbitrary type which could be unrelated to 'unknown'.ts(2322)
configureStore.d.ts(25, 5): The expected type comes from property 'middleware' which is declared here on type 'ConfigureStoreOptions<any, UnknownAction, Tuple<Middlewares<any>>, Tuple<[StoreEnhancer<{ dispatch: {}; }>, StoreEnhancer]>, any>'
```

Should I be using Tuple for this?

**Reply by markerikson** [reply]
Can you show the source code and typing for your `myMiddleware` ?  Also, what version of `redux` do you have installed?

**Reply by devanecondition** [reply]
This is my bad.  I had to update my redux to `5.0.0-beta.0`.  Thanks for the quick reply!

**Reply by spiralabhishek** [reply]
```
export const store: AppStore = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false,
      immutableCheck: false
    }).concat([
      createStateSyncMiddleware({
        channel: "ketchcloth",
        broadcastChannelOption: { type: "localstorage", webWorkerSupport: false },
        blacklist: ['persist/PERSIST', 'persist/PURGE', 'persist/REHYDRATE']
      })
    ])
})
```

I getting same error here 

``
TS2322: Type '(getDefaultMiddleware: GetDefaultMiddleware<any>) => Tuple<[ThunkMiddleware<any, UnknownAction>, ...Middleware<{}, any, Dispatch<AnyAction>>[]]>' is not assignable to type '(getDefaultMiddleware: GetDefaultMiddleware<any>) => Tuple<Middlewares<any>>'.
  Type 'Tuple<[ThunkMiddleware<any, UnknownAction>, ...Middleware<{}, any, Dispatch<AnyAction>>[]]>' is not assignable to type 'Tuple<Middlewares<any>>'.
    Type '[ThunkMiddleware<any, UnknownAction>, ...Middleware<{}, any, Dispatch<AnyAction>>[]]' is not assignable to type 'Middlewares<any>'.
      Type 'ThunkMiddleware<any, UnknownAction> | Middleware<{}, any, Dispatch<AnyAction>>' is not assignable to type 'Middleware<{}, any, Dispatch<UnknownAction>>'.
        Type 'ThunkMiddleware<any, UnknownAction>' is not assignable to type 'Middleware<{}, any, Dispatch<UnknownAction>>'.
          Types of parameters 'next' and 'next' are incompatible.
            Type 'unknown' is not assignable to type 'T'.
              'T' could be instantiated with an arbitrary type which could be unrelated to 'unknown'.
    48 | export const store: AppStore = configureStore({
    49 |   reducer: rootReducer,
  > 50 |   middleware: (getDefaultMiddleware) =>
       |   ^^^^^^^^^^
    51 |     getDefaultMiddleware({
    52 |       serializableCheck: false,
    53 |       immutableCheck: false
    ``

**Reply by spiralabhishek** [reply]
``
"@reduxjs/toolkit": "^2.1.0",
"redux": "^5.0.1",
``

**Reply by spiralabhishek** [reply]
`import { Action, AnyAction, configureStore, Store, ThunkAction } from "@reduxjs/toolkit"
import { useDispatch } from "react-redux"
import { ThunkDispatch } from "redux-thunk"
import { persistStore, persistReducer } from "redux-persist"
import { createStateSyncMiddleware } from "redux-state-sync"
// import storage from "redux-persist/lib/storage"
import { Persistor } from "redux-persist/lib/types"
import { encryptTransform } from "redux-persist-transform-encrypt"
import combineReducer from "redux/reducer"
import localForage from "localforage"

const persistConfig = {
  key: "ketchcloth",
  storage: localForage,
  transforms: [
    encryptTransform({
      secretKey: process.env.REACT_APP_PERSIST_SECRET_KEY || "",
      onError: (error: Error) => {
        // Handle the error.
        console.log("ERROR", error)
      }
    })
  ]
}

export type RootState = ReturnType<typeof combineReducer>

const persistedReducer = persistReducer(persistConfig, combineReducer)

// 2. Create a type for thunk dispatch
export type AppThunkDispatch = ThunkDispatch<RootState, any, AnyAction>

export type AppStore = Omit<Store<RootState, AnyAction>, "dispatch"> & {
  dispatch: AppThunkDispatch
}

const rootReducer = (state: any, action: AnyAction) => {
  if (action.type === "auth/logout/fulfilled") {
    state = {
      ...state,
      Customer: undefined,
      ShoppingCartState: undefined
    }
  }
  return persistedReducer(state, action)
}

export const store: AppStore = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false,
      immutableCheck: false
    }).concat([
      createStateSyncMiddleware({
        channel: "ketchcloth",
        broadcastChannelOption: { type: "localstorage", webWorkerSupport: false },
        blacklist: ['persist/PERSIST', 'persist/PURGE', 'persist/REHYDRATE']
      })
    ])
})

export const persistor: Persistor = persistStore(store)

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>

export const useAppDispatch = () => useDispatch<AppThunkDispatch>()`

**Comment by ghost** [comment]
`combineSlices`

I came up with this 1 year ago :))

Glad it is going to be supported in the main library and I don't have to maintain the codebase for long :D

These were my junk takes on this matter:
```
//Extend slice function with wrapped reducer
/**
 *  @remarks
 * 
 *  This method extends an already declared slice in oreder to attach custom actions to the reducer.
 * 
 *  @param originalSlice The root slice you want to extend
 *  @param caseReducers The case reducers you want to add to the current ones. Current slice reducers will be override.
 *  @param extraReducers Extra reducers to add to the slice.
 *  @param newName The resulted slice new name if you want to override it. If not, it will get the originalSlice name
 *  @param newInitialState The resulted slice initalState. If none is provided, the originalSlice's initalState will be preserved
 * 
 *  @returns slice: Slice
 *  The extended slice containing the current reducers concatenated with the new caseReducers and with the extraReducers 
 *
 */

export const extendSliceWrap = ( originalSlice: Slice, caseReducers?: ValidateSliceCaseReducers<any, SliceCaseReducers<any>>, extraReducers?: ( ( builder: ActionReducerMapBuilder<NoInfer<any>> ) => void ), newName?: string, newInitialState?: any ) => {

  const wrappingSlice = createSlice( {
    name: newName || originalSlice.name,
    initialState: newInitialState || originalSlice.getInitialState(),
    reducers: { ...caseReducers },
    extraReducers
  } );
  const wrappedReducer = ( state: RootState, action: any ) => originalSlice.reducer( wrappingSlice.reducer( state, action ), action );
  return {
    ...wrappingSlice,
    reducer: wrappedReducer,
    caseReducers: {
      ...originalSlice.caseReducers,
      ...wrappingSlice.caseReducers
    }
  };
};
```
```
//Extend slice function
/**
 *  @remarks
 * 
 *  This method extends an already declared slice in oreder to attach custom actions to the reducer.
 * 
 *  @param originalSlice The root slice you want to extend
 *  @param caseReducers The case reducers you want to add to the current ones. Current slice reducers will be override.
 *  @param extraReducers Extra reducers to add to the slice.
 *  @param newName The resulted slice new name if you want to override it. If not, it will get the originalSlice name
 *  @param newInitialState The resulted slice initalState. If none is provided, the originalSlice's initalState will be preserved
 * 
 *  @returns slice: Slice
 *  The extended slice containing the current reducers concatenated with the new caseReducers and with the extraReducers 
 *
 */

export const extendSlice = ( originalSlice: Slice, caseReducers?: ValidateSliceCaseReducers<any, SliceCaseReducers<any>>, extraReducers?: ( ( builder: ActionReducerMapBuilder<NoInfer<any>> ) => void ), newName?: string, newInitialState?: any ) => {
  return createSlice( {
    name: newName || originalSlice.name,
    initialState: newInitialState || originalSlice.getInitialState(),
    reducers: { ...caseReducers, ...originalSlice.caseReducers },
    extraReducers
  } );
};
 ```
 
Hopefully the extrareducers can merge as well.

**Reply by EskiMojo14** [reply]
i think you're misunderstanding what `combineSlices` does - it doesn't allow extending an existing slice like you have there.

it acts like [combineReducers](https://redux.js.org/api/combinereducers), creating a new reducer where each key of the state is managed by a different reducer.

```ts
const countSlice = createSlice({
  name: "count",
  initialState: { value: 0 },
  reducers: {
    increment: (state) => { state.value++ }
  }
});
const combinedReducer = combineSlices(countSlice);
const initialState = combinedReducer(undefined, anAction()); // { count: { value: 0 } }
const nextState = combinedReducer(initialState, countSlice.actions.increment()); // { count: { value: 1 } }
```

We have no plans of any utilities for extending slices like your example. I'd be curious what you're using it for, if I'm honest.

**Reply by EskiMojo14** [reply]
on the off chance that i needed two slices/reducers managing the same part of state, i would likely use [reduce-reducers](https://github.com/redux-utilities/reduce-reducers) to do so:
```ts
const countSlice = createSlice({
  name: "count",
  initialState: { value: 0 },
  reducers: {
    increment: (state) => { state.value++ }
  }
});

const countSlice2 = createSlice({
  name: "count",
  initialState: { value: 0 },
  reducers: {
    decrement: (state) => { state.value-- }
  }
});

export const reducer = reduceReducers(countSlice.reducer, countSlice2.reducer);
```

**Reply by ghost** [reply]
Ok. Is there an official way to combine two slices? Like with the case reducers and extraReducers combined?

**Reply by EskiMojo14** [reply]
no, that's not something we support. could you give an example of where that would be useful?

**Reply by ghost** [reply]
A personal example: I am developing a toolkit where the a slice that has some logic and a client wanted to attach something extra to it. We did not want to implement that change in the codebase because it was quite specific so we added that function that basically combines the two slices together.

**Reply by EskiMojo14** [reply]
then yeah, i would suggest using the approach above with reduce-reducers
it's a lot simpler both to type and functionally

**Reply by ghost** [reply]
Is it a library this reduce-reducers? Sorry if I am dumb founded. I searched in the docs for this reduce reducers and no sign of it.

**Reply by EskiMojo14** [reply]
```ts
import { providedSlice } from "your-package";

const clientSlice = createSlice({
  name: providedSlice.name,
  initialState: providedSlice.getInitialState,
  reducers: {
    extraClientAction: (state) => {}
  }
})

export const { extraClientAction } = clientSlice.actions;

export const finalReducer = reduceReducers(providedSlice.reducer, clientSlice.reducer);
```

**Reply by EskiMojo14** [reply]
> Is it a library this reduce-reducers? Sorry if I am dumb founded. I searched in the docs for this reduce reducers and no sign of it.

yes, i linked it in my previous comment

**Reply by EskiMojo14** [reply]
it's also mentioned in the FAQ: https://redux.js.org/faq/reducers#how-do-i-share-state-between-two-reducers-do-i-have-to-use-combinereducers
